### PR TITLE
Tajul/keyword async processing

### LINKED ISSUE #8: [Chore] Brittle asynchronous processing of keywords
## Issue

Upon uploading keywords, the scraping will be performed asynchronously via an event listener

https://github.com/tajul-saajan/google-searcher/blob/36a95d4f0ee6216e9fb4abdb15038b12d2f25d54/src/search/listeners/search.listener.ts#L17-L27

All keywords are processed in a loop. If any error occurs in the loop, the latter will stop, and all remaining keywords will not be scraped and will not even have been saved to the database. A queue (supported by Nest.JS, see: https://docs.nestjs.com/techniques/queues) would be a better fit in this case.

> [!WARNING]
> Using the observer pattern can be beneficial when there is more than one subscriber/consumer. Since the event is consumed in the same module as it is emitted, it creates indirection unnecessarily. 

## Expected

- Keywords should be stored in the database with a flag to track the scraping status as soon as they are uploaded 
- A distinct background job to scrap the Google page for each keyword is then scheduled.

The benefits are:

- Fast CSV upload request
- The scraping of each keyword is isolated, i.e., one could error while others could be successful.

### PR DESCRIPTION
**Pull request by tajul-saajan**
resolves issue #8 

- Added field boolean isProcessed in searchStat entity
- In entity, only keyword and isProcessed are required, all other fields are optional
- After parsing the file, save it to the database with the keyword and then add the record to the queue
- Added queue consumer which performs a search using searcher and then updates the records to the database with the search result
- Added support for job failure, configured queue to have attempts and backoff time

### CODE REVIEW

File: src/entities/search-stat.entity.ts [hunk 1]
```diff
@@ -14,18 +14,21 @@ export class SearchStat {
...
-  @Column({ type: 'bigint' })
+  @Column({ type: 'bigint', nullable: true })
   totalResultsCount: number;
 
-  @Column({ type: 'text' })
+  @Column({ type: 'text', nullable: true })
   cachedResponse: string;
 
+  @Column({ type: 'bool', default: false })
+  isProcessed: boolean;
```
Comment (olivierobert) [inline src/entities/search-stat.entity.ts:30]: This does not account for a failed status. How about instead using an enum:

``` 
enum SearchStatus {
    Pending = 'pending',
    Completed = 'completed',
    Failed = 'failed'
}

@Entity()
export class SearchStat {
  ...
  @Column({ type: 'text' })
  status: SearchStatus;
}
```

Comment (tajul-saajan) [reply src/entities/search-stat.entity.ts:30]: I did not consider the failed status, as in the queue used max attempts and exponential backoff delay to make sure the failed job retries continue with exponential backoff.  But there can be cases, the job exhausts all the retries and unable to process the search. Should have considered this scenerio

Comment (tajul-saajan) [reply src/entities/search-stat.entity.ts:30]: In that case, when the queued job exhausts all its attempts and is unable to process the search operation, I guess in that scenario I need to add a cron job scheduler  to  pick all failed jobs and process them again?

**Comment by tajul-saajan** [comment]
In the above commit 
- status field added in search stat entity
- added status change based on queued job success or failure

File: src/entities/search-stat.entity.ts (hunk 1 above)
Comment (olivierobert) [reply src/entities/search-stat.entity.ts:31]: In a production application, taking care of the failed jobs could take many forms (retrying them, or sending a report/error), but it is outside of the scope of this exercise.

As for the exponential backoff, still, the status could be marked as failed in the mean time to differentiate it from other jobs 💡

**Pull request merged**